sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -280,9 +297,15 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
         rule.applyOrElse(this, identity[BaseType])
```

Comment:
Here after applying the `rule`, we need to carry over the tags, if needed, too.

Since carrying the tags in both if and else branches, so maybe:

```scala
val newNode = if (this fastEquals afterRuleOnChildren) {
  CurrentOrigin.withOrigin(origin) {
    rule.applyOrElse(this, identity[BaseType])
  }
} else {
  CurrentOrigin.withOrigin(origin) {
    rule.applyOrElse(afterRuleOnChildren, identity[BaseType])
  }
}

// carrying over the tags to newNode...
```